File: TopologicalSort/269. Alien Dictionary/269. Alien Dictionary.cpp
class Solution {
public:
    string alienOrder(vector<string>& words) {
        //if (words == NULL || words.size() == 0) return "";
        unordered_map<char, unordered_set<char>> next;
        unordered_map<char, int> inDegree;
        for (auto w : words) {
            for (auto c : w) inDegree[c] = 0;
        }
        
        for (int i = 0; i < words.size()-1; i++) {
            string s = words[i];
            string t = words[i+1];
            if (s.size() > t.size() && s.substr(0, t.size()) == t) return "";
            
            for (int j = 0; j < min(s.size(), t.size()); j++) {
                if (s[j] == t[j]) continue;
                if (next[s[j]].find(t[j]) == next[s[j]].end()) { // first different place will exit
                    next[s[j]].insert(t[j]);
                    inDegree[t[j]]++;
                }
                break;
            }
        }
        
        // topology sort(BFS)
        queue<char> q;
        
        for (auto x : inDegree) {
            if (x.second == 0)  q.push(x.first);
        }
        
        string ret;
        while (!q.empty()) {
            char cur = q.front();
            q.pop();
            ret.push_back(cur);
            
            for (auto c : next[cur]) {
                inDegree[c]--;
                if (inDegree[c] == 0) q.push(c);
            }
        }
        
        return ret.size() == inDegree.size() ? ret : "";
    }
};

File: TopologicalSort/269. Alien Dictionary/269. Alien Dictionary.java
class Solution {
    // create dag then topology sort
    public String alienOrder(String[] words) {
        if (words == null || words.length == 0) return "";
        Map<Character, Set<Character>> map = new HashMap<>();
        Set<Character> set = new HashSet<>();
        for (String w : words) {
            for (char c : w.toCharArray()) set.add(c);
        }
        int[] indegree = new int[26];
        for (int i = 0; i < words.length-1; i++) {
            String s1 = words[i];
            String s2 = words[i+1];
            for (int j = 0; j < s1.length(); j++) {
                if (j >= s2.length()) return "";
                if (s1.charAt(j)  != s2.charAt(j)) { // first different place will exit
                    if (map.containsKey(s1.charAt(j)) && map.get(s1.charAt(j)).contains(s2.charAt(j))) break;
                    map.computeIfAbsent(s1.charAt(j), k-> new HashSet<>()).add(s2.charAt(j));
                    indegree[s2.charAt(j)-'a']++;
                    break;
                }
            }
        }
        
        // topology sort(BFS)
        Queue<Character> q = new LinkedList<>();
        StringBuilder sb = new StringBuilder();
        for (char c : set) {
            if (set.contains(c) && indegree[c-'a'] == 0)  {
                sb.append(c);
                q.offer(c);
            }
        }
        
        while (!q.isEmpty()) {
            char c = q.poll();
            if (!map.containsKey(c)) continue;
            for (char next : map.get(c)) {
                indegree[next-'a']--;
                if (set.contains(next) && indegree[next-'a'] == 0) {
                    sb.append(next);
                    q.offer(next);
                }
            }
        }
        
        return sb.length() == set.size() ? sb.toString() : "";
    }
}
